packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts

```diff
@@ -14,46 +15,77 @@ import { useGridSelector } from '../features/core/useGridSelector';
 import { useGridState } from '../features/core/useGridState';
 import { gridDensityRowHeightSelector } from '../features/density/densitySelector';
 import { visibleGridRowCountSelector } from '../features/filter/gridFilterSelector';
-import { gridPaginationSelector } from '../features/pagination/gridPaginationSelector';
+import {
+  gridPageSizeSelector,
+  gridPageSelector,
+} from '../features/pagination/gridPaginationSelector';
 import { optionsSelector } from '../utils/optionsSelector';
 import { useLogger } from '../utils/useLogger';
 import { useGridApiEventHandler } from './useGridApiEventHandler';
+import { GridComponentProps } from '../../GridComponentProps';
+import { getBrowserScrollbarSize } from '../../utils';
 
-export const useGridContainerProps = (apiRef: GridApiRef) => {
+export const useGridContainerProps = (
+  apiRef: GridApiRef,
+  props: Pick<GridComponentProps, 'scrollbarSize'>,
+) => {
   const logger = useLogger('useGridContainerProps');
   const [gridState, setGridState, forceUpdate] = useGridState(apiRef);
   const windowSizesRef = React.useRef<ElementSize>({ width: 0, height: 0 });
   const options = useGridSelector(apiRef, optionsSelector);
   const rowHeight = useGridSelector(apiRef, gridDensityRowHeightSelector);
   const columnsTotalWidth = useGridSelector(apiRef, gridColumnsTotalWidthSelector);
   const visibleRowsCount = useGridSelector(apiRef, visibleGridRowCountSelector);
-  const paginationState = useGridSelector(apiRef, gridPaginationSelector);
+  const pageState = useGridSelector(apiRef, gridPageSelector);
+  const pageSizeState = useGridSelector(apiRef, gridPageSizeSelector);
   const windowRef = apiRef.current.windowRef;
 
   const getVirtualRowCount = React.useCallback(() => {
     logger.debug('Calculating virtual row count.');
-    if (options.pagination && !options.autoPageSize) {
-      const rowsLeft = visibleRowsCount - paginationState.page * paginationState.pageSize;
-      return rowsLeft > paginationState.pageSize ? paginationState.pageSize : rowsLeft;
+    if (options.pagination && (!options.autoPageSize || options.pageSize)) {
+      const rowsLeft = visibleRowsCount - pageState.currentPage * pageSizeState;
+      return rowsLeft > pageSizeState ? pageSizeState : rowsLeft;
     }
     return visibleRowsCount;
   }, [
     logger,
     options.autoPageSize,
     options.pagination,
-    paginationState.page,
-    paginationState.pageSize,
+    options.pageSize,
+    pageState.currentPage,
+    pageSizeState,
     visibleRowsCount,
   ]);
 
+  const getScrollbarSize = React.useCallback(() => {
```

Comment:
we should already have one

Reply:
Their was one in `useGridScrollbarSizeDetector.tsx` that was updating the `options.scrollbarWidth`
But now that I don't use the option, I need to store it elsewhere.

It is only used in `useGridContainerProps.ts` so call it directly here.
I would store it in a `state.scrollbarWidth` but I don't see a lot of benefits to do it.

Reply:
well why don't you use options.scrollbarWidth, we don't need to stop using options completely, we can just use it when it is useful.

Reply:
For me the option pattern makes the code quite hard to follow and to debug.

You have optional props
Then a hook `useOptionsProp` that put those props in the state and sometime applies a default value, sometime merge the value with a default value (for `classes` and `localeText`) and sometime does not apply any default value.
Then a hook `useGridScrollbarSizeDetector` that updates the `scrollbarSize` option but only if the prop is null.

Reply:
I understand your confusion and options should be refactored a bit now. However I think in the case of scrollbarSize it is still useful, you could refactor useGridScrollbarSizeDetector to its own state but I think it's a bit outside this PR scope.

Reply:
I can go back using the option.
When an effect uses both options and props, they may be inconsistencies when the user updates 2 props at once (their is a render where the prop used directly is updated but the prop used through the option process is not).

But I suppose for the scrollbarWidth, it should not occurs.

Reply:
For me, all the "options" that can be initialized synchronously before the Grid DOM is mounted should be preprocessed in some `useMemo`s

Something like 

```ts
const props = useGridPreprocessedProps(rawProps)
```

And most of them are. Their is only the `scrollbarSize` which is a bit different since you pass the grid root (I suppose they may be situations where the scrollwidth is not the same in various places of the DOM).
So if in a different PR I put the `scrollbarbarWidth` out of the options (in a dedicated substate or in `state.scrollbar`), then I think we can have every use case currently covered by the options without the prop to state pattern.

Reply:
Props are already merged with their default value. 
https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/GridComponent.tsx#L38

In the case of the scrollbar, this can't work as the default value is dynamic hence why we need to use options at this stage.

Reply:
Yes, my proposal allow to stop using the option (or any state-based solution) for `classes` and `localeText`.
But the the scrollbarWidth we need the grid to be mounted to determine the default value.